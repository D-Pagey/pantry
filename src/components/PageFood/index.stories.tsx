import React from 'react';

import { UserDan, Fridge, TenantHeidi } from '../../fixtures';
import { AuthContext } from '../ProviderAuth';
import { PageFood } from '.';

export default { title: 'PageFood' };

export const normal = () => (
    <AuthContext.Provider
        value={{
            fetchUserData: () => {},
            isAuthed: true,
            isCheckingAuth: false,
            setUser: () => {},
            signOut: () => {},
            user: UserDan
        }}
    >
        <PageFood fridge={Fridge} tenants={[TenantHeidi]} />
    </AuthContext.Provider>
);
