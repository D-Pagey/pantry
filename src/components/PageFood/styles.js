import styled from 'styled-components';
import { Link } from 'react-router-dom';

import { Button as Btn } from '../Button';

export const Wrapper = styled.div``;

export const RouterLink = styled(Link)`
    color: inherit;
    text-decoration: none;
`;

export const FoodCardWrapper = styled.div`
    align-items: center;
    display: flex;
    flex-direction: column;
`;

export const Button = styled(Btn)`
    justify-self: end;
    margin: 0 1rem;
`;
