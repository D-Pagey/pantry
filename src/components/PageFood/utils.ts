import { DatabaseCategoryType, FoodTypes } from '../../types';

export const filterFridge = (fridge: FoodTypes[], category: DatabaseCategoryType): FoodTypes[] => {
    return fridge.filter((food) => {
        if (food.categories.includes(category.id)) {
            return true;
        }

        return false;
    });
};


// TODO: Refactor this to be one function
export const swapIdsForNames = (fridge: FoodTypes[], categories: DatabaseCategoryType[]): FoodTypes[] => {
    return fridge.map(foodItem => {
        const categoryNames = [...foodItem.categories.map(categoryId => {
            const fullCategory = categories.filter(item => item.id === categoryId);
            return fullCategory[0].name;
        })];
        
        return {
            ...foodItem,
            categories: categoryNames
        };
    });
};

export const swapNamesForIds = (fridge: FoodTypes[], categories: DatabaseCategoryType[]): FoodTypes[] => {
    return fridge.map(foodItem => {
        const categoryIds = [...foodItem.categories.map(categoryName => {
            const fullCategory = categories.filter(item => item.name === categoryName);
            return fullCategory[0].id;
        })];
        
        return {
            ...foodItem,
            categories: categoryIds
        };
    });
};