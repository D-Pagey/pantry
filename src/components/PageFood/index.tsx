import React, { FC, useCallback, useContext, useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useMediaQuery } from 'react-responsive';
import arraySort from 'array-sort';
import { toast } from 'react-toastify';

import { db } from '../../services';
import { FoodType, TenantType } from '../../types';
import { getExpiringItems, filterFridgeByCategory, getCategoriesAndCounts, filterByTenantIds } from '../../utils';
import { mediaQuery } from '../../tokens';
import { Layout } from '../Layout';
import { CategoryFilterMobile } from '../CategoryFilterMobile';
import { CategoryFilterDesktop } from '../CategoryFilterDesktop';
import { OwnerFilter } from '../OwnerFilter';
import { FoodCard } from '../FoodCard';
import { FoodOptions } from '../FoodOptions';
import { AuthContext } from '../ProviderAuth';
import * as S from './styles';

type PageFoodProps = {
    fridge: FoodType[];
    tenants: TenantType[];
};

export const PageFood: FC<PageFoodProps> = ({ fridge, tenants }) => {
    const [selectedFood, setSelectedFood] = useState<FoodType[]>();
    const [category, setCategory] = useState('all');
    const [isExpiring, setIsExpiring] = useState(false);
    const [editingItem, setEditingItem] = useState<FoodType | undefined>();
    const [selectedTenants, setSelectedTenants] = useState<string[]>([]);
    const { user } = useContext(AuthContext);
    const history = useHistory();
    const isTabletOrLarger = useMediaQuery({
        query: mediaQuery.tablet
    });

    const deleteFoodItem = (name: string): void => {
        if (user) {
            db.collection('households')
                .doc(user.household)
                .update({ [`fridge.${name}.batches`]: {} })
                .then(() => toast.error('Food deleted'))
                .catch(() => toast.error('Error with deleting food'));
        }
    };

    const filterFood = useCallback(() => {
        const isCategoryAll = category === 'all';

        let filteredFridge = [...fridge];

        if (!isCategoryAll) {
            filteredFridge = filterFridgeByCategory(filteredFridge, category);
        }

        if (isExpiring) {
            filteredFridge = getExpiringItems(filteredFridge);
        }

        if (selectedTenants.length > 0) {
            filteredFridge = filterByTenantIds(filteredFridge, selectedTenants);
        }

        return filteredFridge;
    }, [category, fridge, isExpiring, selectedTenants]);

    useEffect(() => {
        if (fridge.length > 0 && tenants.length > 0) {
            setSelectedFood(filterFood());
        }
    }, [category, selectedTenants, isExpiring, filterFood, fridge, tenants]);

    const handleCategoryClick = (selectedCategory: string): void => {
        setCategory(selectedCategory);
    };

    const handleExpiringClick = (): void => {
        setIsExpiring(!isExpiring);
    };

    const handleFoodDelete = (): void => {
        if (editingItem?.name) {
            deleteFoodItem(editingItem.name);
            setEditingItem(undefined);
        }
    };

    const handleFoodEdit = (): void => {
        history.push(`/${editingItem?.name}/edit`);
    };

    const handleFoodClick = (item: FoodType) => (): void => {
        if (!editingItem || editingItem.name !== item.name) setEditingItem(item);
        if (editingItem?.name === item.name) setEditingItem(undefined);
    };

    const handleAddClick = () => history.push('/add');

    return (
        <Layout title="Your Food:">
            <S.Wrapper>
                {isTabletOrLarger ? (
                    <CategoryFilterDesktop
                        categories={getCategoriesAndCounts(fridge)}
                        selected={category}
                        handleCategoryClick={handleCategoryClick}
                    />
                ) : (
                    <CategoryFilterMobile selected={category} setSelected={handleCategoryClick} />
                )}

                <S.FilterWrapper>
                    <OwnerFilter
                        tenants={tenants}
                        setSelectedTenants={setSelectedTenants}
                        selectedTenants={selectedTenants}
                    />

                    <S.ExpiringButton handleClick={handleExpiringClick} isEnabled={isExpiring} />
                </S.FilterWrapper>

                {fridge?.length === 0 && <p data-testid="pageFoodNoData">You have no food in your fridge.</p>}

                {fridge?.length !== 0 && selectedFood?.length === 0 && (
                    <p data-testid={`pageFoodNoData${category}`}>
                        There is no {isExpiring && 'expiring'} food that falls under the category of {category}
                    </p>
                )}

                <S.FoodCardGrid>
                    {selectedFood &&
                        arraySort(selectedFood, 'name').map((item: FoodType) => {
                            if (item.batches.length > 0) {
                                return (
                                    <FoodCard
                                        handleClick={handleFoodClick(item)}
                                        isSelected={item.name === editingItem?.name}
                                        item={item}
                                        key={item.name}
                                        margin="0 0 1rem"
                                        tenants={tenants}
                                    />
                                );
                            }

                            return null;
                        })}
                </S.FoodCardGrid>

                <S.AddButton size="sm" onClick={handleAddClick}>
                    Add Item
                </S.AddButton>
            </S.Wrapper>

            {editingItem && (
                <FoodOptions name={editingItem.name} handleDelete={handleFoodDelete} handleEdit={handleFoodEdit} />
            )}
        </Layout>
    );
};
