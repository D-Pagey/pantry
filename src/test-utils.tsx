import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import React, { FC, ReactNode } from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { FirebaseContext } from './components/ProviderFirebase';

type AllTheProvidersProps = {
    children: ReactNode;
    customContext?: any;
};

const AllTheProviders: FC<AllTheProvidersProps> = ({ children, customContext }) => (
    <FirebaseContext.Provider
        value={{
            categories: [],
            deleteFoodItem: () => () => null,
            expiringCount: 0,
            isAuthed: false,
            isCheckingAuth: false,
            fridge: [],
            signOut: () => null,
            updateFridge: () => null,
            user: {
                email: null,
                name: null
            },
            ...customContext
        }}
    >
        <MemoryRouter>{children}</MemoryRouter>
    </FirebaseContext.Provider>
);

const customRender = (ui: any, customContext?: any, options?: any) =>
    render(ui, { wrapper: () => <AllTheProviders customContext={customContext}>{ui}</AllTheProviders>, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
