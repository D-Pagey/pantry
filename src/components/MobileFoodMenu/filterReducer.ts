import { SortOptions } from '../PageFood/foodReducer';

type ChangeSortAction = {
    type: 'CHANGE_SORTED_BY';
    sortBy: SortOptions;
};

type ChangeShowExpiredAction = {
    type: 'CHANGE_SHOW_ONLY_EXPIRED';
    onlyShowExpired: boolean;
};

type ToggleSelectedOwnerAction = {
    type: 'TOGGLE_SELECTED_OWNER';
    ownerId: string;
};

type ResetAction = {
    type: 'RESET';
    foodPageFilters: FilterState;
};

type ChangeCategoryAction = {
    type: 'CHANGE_CATEGORY';
    category: string;
};

export type FilterState = {
    selectedOwners: string[];
    showOnlyExpiring: boolean;
    sortBy: SortOptions;
    category: string;
};

type FilterActions =
    | ChangeSortAction
    | ChangeShowExpiredAction
    | ToggleSelectedOwnerAction
    | ResetAction
    | ChangeCategoryAction;

export const init = (initialFilterState: FilterState, foodPageFilters: FilterState): FilterState => {
    return {
        ...initialFilterState,
        ...foodPageFilters
    };
};

export const initialFilterState: FilterState = {
    selectedOwners: [],
    showOnlyExpiring: false,
    sortBy: 'date',
    category: ''
};

export const filterReducer = (state: FilterState, action: FilterActions): FilterState => {
    switch (action.type) {
        case 'CHANGE_SORTED_BY': {
            return {
                ...state,
                sortBy: action.sortBy
            };
        }

        case 'CHANGE_SHOW_ONLY_EXPIRED': {
            return {
                ...state,
                showOnlyExpiring: action.onlyShowExpired
            };
        }

        case 'TOGGLE_SELECTED_OWNER': {
            if (state.selectedOwners.includes(action.ownerId)) {
                return {
                    ...state,
                    selectedOwners: state.selectedOwners.filter((id) => id !== action.ownerId)
                };
            } else {
                return {
                    ...state,
                    selectedOwners: [...state.selectedOwners, action.ownerId]
                };
            }
        }

        case 'RESET': {
            return init(initialFilterState, action.foodPageFilters);
        }

        case 'CHANGE_CATEGORY': {
            return {
                ...state,
                category: action.category
            };
        }
    }
};
