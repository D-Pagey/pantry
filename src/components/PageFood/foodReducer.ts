import { FoodType, TenantType } from '../../types';
import { getExpiringItems } from '../../utils';
import { FilterState } from '../MobileFoodMenu/filterReducer';
import { applyMultipleFilters } from './utils';

export type SortOptions = 'name' | 'date';

type ApplyFiltersAction = {
    type: 'APPLY_FILTERS';
    filters: FilterState;
    fridge: FoodType[];
};

type RemoveExpiringFilterAction = {
    type: 'REMOVE_EXPIRING_FILTER';
    fridge: FoodType[];
};

type RemoveSelectedOwnersAction = {
    type: 'REMOVE_SELECTED_OWNERS';
    fridge: FoodType[];
    tenants: TenantType[];
};

type UpdateFridgeAction = {
    type: 'UPDATE_FRIDGE';
    fridge: FoodType[];
};

type FoodActions = ApplyFiltersAction | RemoveExpiringFilterAction | UpdateFridgeAction | RemoveSelectedOwnersAction;

export type FoodState = {
    food: FoodType[];
    filters: FilterState;
};

export const init = (initialFoodState: FoodState, tenants: TenantType[], fridge: FoodType[]): FoodState => {
    const expiringItems = getExpiringItems(fridge);

    const filters = {
        ...initialFoodState.filters,
        showOnlyExpiring: expiringItems.length > 0,
        selectedOwners: tenants.map((tenant) => tenant.uid)
    };

    return {
        ...initialFoodState,
        filters,
        food: applyMultipleFilters(fridge, filters)
    };
};

export const initialFoodState: FoodState = {
    food: [],
    filters: {
        selectedOwners: [],
        showOnlyExpiring: true,
        sortBy: 'date'
    }
};

export const foodReducer = (state: FoodState, action: FoodActions): FoodState => {
    switch (action.type) {
        case 'APPLY_FILTERS': {
            return {
                ...state,
                filters: action.filters,
                food: applyMultipleFilters(action.fridge, action.filters)
            };
        }

        case 'REMOVE_EXPIRING_FILTER': {
            const updatedFilters = {
                ...state.filters,
                showOnlyExpiring: false
            };

            return {
                ...state,
                filters: updatedFilters,
                food: applyMultipleFilters(action.fridge, updatedFilters)
            };
        }

        case 'REMOVE_SELECTED_OWNERS': {
            const updatedFilters = {
                ...state.filters,
                selectedOwners: action.tenants.map((tenant) => tenant.uid)
            };

            return {
                ...state,
                filters: updatedFilters,
                food: applyMultipleFilters(action.fridge, updatedFilters)
            };
        }

        case 'UPDATE_FRIDGE': {
            return {
                ...state,
                food: applyMultipleFilters(action.fridge, state.filters)
            };
        }

        default:
            return state;
    }
};
