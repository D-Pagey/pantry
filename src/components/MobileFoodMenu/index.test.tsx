import userEvent from '@testing-library/user-event';
import ReactModal from 'react-modal';
import { render, screen } from '../../test-utils';
import { Categories, TenantDan, TenantHeidi } from '../../fixtures';
import { SortOptions } from '../PageFood/foodReducer';
import { FilterState } from './filterReducer';
import { MobileFoodMenu } from '.';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    // @ts-ignore
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockHistoryPush
    })
}));

const props = {
    categories: Categories,
    handleFoodDelete: () => null,
    handleApplyFilters: () => null,
    tenants: [TenantHeidi, TenantDan],
    foodPageFilters: {
        selectedOwners: [TenantHeidi.uid, TenantDan.uid],
        showOnlyExpiring: true,
        sortBy: 'name' as SortOptions,
        category: ''
    }
};

describe('MobileFoodMenu component', () => {
    it('should render', () => {
        const { container } = render(<MobileFoodMenu {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should open modal when filter button clicked', () => {
        const { container } = render(<MobileFoodMenu {...props} />);
        ReactModal.setAppElement(container);

        userEvent.click(screen.getByTestId('filterMenuButton'));

        screen.getByText('Set Filters');
    });

    it('should call handleFoodDelete if delete button clicked', () => {
        const handleFoodDelete = jest.fn();

        render(<MobileFoodMenu {...props} handleFoodDelete={handleFoodDelete} editingItemName="chicken" />);

        userEvent.click(screen.getByTestId('mobileFoodMenuDeleteButton'));

        expect(handleFoodDelete).toHaveBeenCalled();
    });

    it('should call history push if edit button clicked', () => {
        const editingItemName = 'chicken';

        render(<MobileFoodMenu {...props} editingItemName={editingItemName} />);

        userEvent.click(screen.getByTestId('mobileFoodMenuEditButton'));

        expect(mockHistoryPush).toHaveBeenCalledWith(`/${editingItemName}/edit`);
    });

    it('should call history push if add button clicked', () => {
        render(<MobileFoodMenu {...props} />);

        userEvent.click(screen.getByText('Add Item'));

        expect(mockHistoryPush).toHaveBeenCalledWith('/add');
    });

    it('should call handleApplyFilters with changed values', async () => {
        const handleApplyFilters = jest.fn();

        const { container } = render(<MobileFoodMenu {...props} handleApplyFilters={handleApplyFilters} />);
        ReactModal.setAppElement(container);

        userEvent.click(screen.getByTestId('filterMenuButton'));
        // deselect one tenant in filters
        userEvent.click(screen.getByTestId(`photo-${props.tenants[1].uid}`));
        // click All Items instead of Expiring Soon
        userEvent.click(screen.getByText('All Items'));
        // click Date to sort by date instead of default name
        userEvent.click(screen.getByText('Date'));
        userEvent.click(screen.getByText('Apply Filters'));

        const updatedFilters: FilterState = {
            selectedOwners: [props.tenants[0].uid],
            showOnlyExpiring: false,
            sortBy: 'date',
            category: ''
        };

        expect(handleApplyFilters).toHaveBeenCalledWith(updatedFilters);
    });

    it('add an owner if not selected', () => {
        const handleApplyFilters = jest.fn();
        const foodPageFilters = {
            ...props.foodPageFilters,
            selectedOwners: []
        };
        const ownerToSelect = props.tenants[1].uid;

        const { container } = render(
            <MobileFoodMenu {...props} handleApplyFilters={handleApplyFilters} foodPageFilters={foodPageFilters} />
        );
        ReactModal.setAppElement(container);

        userEvent.click(screen.getByTestId('filterMenuButton'));
        // select one tenant in filters
        userEvent.click(screen.getByTestId(`photo-${ownerToSelect}`));
        userEvent.click(screen.getByText('Apply Filters'));

        const updatedFilters: FilterState = {
            ...props.foodPageFilters,
            selectedOwners: [ownerToSelect]
        };

        expect(handleApplyFilters).toHaveBeenCalledWith(updatedFilters);
    });

    it('should handle a cancel click', () => {
        const handleApplyFilters = jest.fn();

        const { container } = render(<MobileFoodMenu {...props} handleApplyFilters={handleApplyFilters} />);
        ReactModal.setAppElement(container);

        userEvent.click(screen.getByTestId('filterMenuButton'));
        // deselect one tenant in filters
        userEvent.click(screen.getByTestId(`photo-${props.tenants[1].uid}`));
        // click All Items instead of Expiring Soon
        userEvent.click(screen.getByText('All Items'));
        // click Date to sort by date instead of default name
        userEvent.click(screen.getByText('Date'));

        // Click cancel instead of applying filters
        userEvent.click(screen.getByText('Cancel'));

        // expect modal to have closed
        expect(screen.queryByText('Set Filters')).not.toBeInTheDocument();

        // reopen modal
        userEvent.click(screen.getByTestId('filterMenuButton'));
        // click apply
        userEvent.click(screen.getByText('Apply Filters'));
        // expect state to be same as originally passed in
        expect(handleApplyFilters).toHaveBeenCalledWith(props.foodPageFilters);
    });
});
