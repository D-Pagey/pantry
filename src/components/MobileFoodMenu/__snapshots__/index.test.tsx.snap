// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileFoodMenu component should render 1`] = `
.c3 {
  background-color: #348841;
  border: 0;
  border-radius: 5px;
  color: white;
  cursor: pointer;
  padding: 1rem;
  min-width: 68px;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: white;
  border-top: 1px solid #C2C2C2;
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  left: 0;
  padding: 1rem;
  position: fixed;
  right: 0;
  z-index: 1;
}

.c1 {
  border: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  margin: 0 1rem 0 0;
  width: 50px;
}

.c2 {
  width: 100%;
}

<div
  class="c0"
>
  <button
    class="c1"
    data-testid="filterMenuButton"
    type="button"
  >
    <img
      alt="filter menu"
      class="c2"
      src="filter.svg"
    />
  </button>
  <button
    class="c3"
    type="button"
  >
    Add Item
  </button>
</div>
`;
